apt-hook/json-hook.cc

```diff
@@ -0,0 +1,302 @@
+                result.esm_apps += 1;
+            } else if (version_from_origin_and_archive_ends_with(install, "UbuntuESM", "-infra-security")) {
+                result.esm_infra += 1;
+            } else if (version_from_origin_and_archive_ends_with(install, "Ubuntu", "-security")) {
+                result.standard += 1;
+            }
+        }
+    }
+
+    return true;
+}
+
+void print_counts(security_package_counts &counts) {
+    std::vector<std::string> count_msgs;
+    bool wrote_security = false;
+
+    if (counts.standard + counts.esm_infra + counts.esm_apps == 0) {
+        return;
+    }
+
+    if (counts.standard > 0) {
+        std::stringstream ss;
+        ss << counts.standard << " standard security update";
+        if (counts.standard != 1) {
+            ss << "s";
+        }
+        count_msgs.push_back(ss.str());
+        wrote_security = true;
+    }
+    if (counts.esm_infra > 0) {
+        std::stringstream ss;
+        ss << counts.esm_infra << " esm-infra ";
+        if (!wrote_security) {
+            ss << "security ";
+        }
+        ss << "update";
+        if (counts.esm_infra != 1) {
+            ss << "s";
+        }
+        count_msgs.push_back(ss.str());
+        wrote_security = true;
+    }
+    if (counts.esm_apps > 0) {
+        std::stringstream ss;
+        ss << counts.esm_apps << " esm-apps ";
+        if (!wrote_security) {
+            ss << "security ";
+        }
+        ss << "update";
+        if (counts.esm_apps != 1) {
+            ss << "s";
+        }
+        count_msgs.push_back(ss.str());
+    }
+
+    for (uint i = 0; i < count_msgs.size(); i++) {
+        if (count_msgs.size() == 3) {
+            if (i == 1) {
+                std::cout << ", ";
+            } else if (i == 2) {
+                std::cout << ", and ";
+            }
+        } else if (count_msgs.size() == 2) {
+            if (i == 1) {
+                std::cout << " and ";
+            }
+        }
+        std::cout << count_msgs[i];
+    }
+    std::cout << std::endl;
+}
+
+int main(int argc, char *argv[])
+{
+    (void) argc;
+    (void) argv;
+
+    char *fd_c_str = getenv("APT_HOOK_SOCKET");
+    if (fd_c_str == NULL) {
+       return 0;
+    }
+    std::string fd_str(fd_c_str);
+    if (fd_str == "") {
+       return 0;
+    }
+    int fd = stoi(fd_str);
+    __gnu_cxx::stdio_filebuf<char> apt_socket_in(fd, std::ios::in);
+    __gnu_cxx::stdio_filebuf<char> apt_socket_out(fd, std::ios::out);
+    std::istream socket_in(&apt_socket_in);
+    std::ostream socket_out(&apt_socket_out);
+
+
+    bool success = false;
+
+    // Read hello message, verify version, and get jsonrpc id
+    jsonrpc_request hello_req;
+    success = read_jsonrpc_request(socket_in, hello_req);
+    if (!success) {
+        return 0;
+    }
+    if (hello_req.method != "org.debian.apt.hooks.hello" || hello_req.notification) {
```

Comment:
Why do we need the `OR` clause here ?

Reply:
https://www.jsonrpc.org/specification#notification

A JSONRPC notification is a message without an id. The hello method should not be a notification because we need to respond to it with the corresponding id.